articles/android-application-widgets-tutorial-in-java/index.md

```diff
@@ -0,0 +1,144 @@
+- `Minimum width` and `Minimum height` define the minimum number of cells that the widget can occupy.
+- The `Resizable` option specifies whether the widget can be resized.
+- If the widget requires a ConfigureActivity, select the `Configurations Screen` option.
+
+From `step 2` above, three different files are created:
+- `NewAppWidget.java`: This is the widget class created under the package name `com.demo.widgetsdemo`.
+- `new_app_widget_info.xml`: This is an XML file under the XML folder. It contains widget configurations and can be edited.
+- `new_app_widget.xml`: This is the layout file for the widget located in the layout folder.
+
+**Step 3: Customize widget layout file**
+Open `new_app_widget_info.xml` XML file. In this file, you can add other views that you would like to be part of your widget. Add a button below the already existing text view, as shown below.
+
+```XML
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="?attr/appWidgetBackgroundColor"
+    android:padding="@dimen/widget_margin"
+    android:orientation="vertical"
+    android:theme="@style/ThemeOverlay.AppWidgetsDemo.AppWidgetContainer">
+
+    <TextView
+        android:id="@+id/appwidget_text"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_margin="8dp"
+        android:textAlignment="center"
+        android:background="?attr/appWidgetBackgroundColor"
+        android:contentDescription="DEMO WIDGET"
+        android:text="DEMO WIDGET"
+        android:textColor="?attr/appWidgetTextColor"
+        android:textSize="24sp"
+        android:textStyle="bold|italic" />
+    <Button
+        android:id="@+id/button1"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_margin="8dp"
+        android:background="?attr/appWidgetBackgroundColor"
+        android:text="CLICK BUTTON"
+        android:textSize="24sp"
+        android:textStyle="bold|italic" />
+</LinearLayout>
+```
+
+**Step 4: Widget Java class**
+Open the `NewAppWidget.java` file. This class extends with the `AppWidgetProvider` class. `AppWidgetProvider` receives and handles broadcast events relevant to the App Widget. `AppWidgetProvider` class overrides the following methods.
+
+- `onUpdate()`: This method updates App Widgets on intervals defined in the `AppWidgetProviderInfo Metadata`. In this case, the `AppWidgetProviderInfo Metadata` is defined in the file `new_app_widget_info.xml`. This method is also called when the App Widget is dropped on the home screen.
+- `onAppWidgetOptionsChanged()`: This method is called when the widget is resized.
+- `onEnabled()`: This method is called when a widget is created.
+`onDisabled()`: This method is called when the widget's last instance is deleted. For example, we can launch an activity within this function.
```

Comment:
I think you can have the java code here just like how you've displayed your XML code.

Reply:
I have seen the picture. I was wondering if we could have the code snippet instead. It's much better than a picture.